sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java

```diff
@@ -202,6 +261,220 @@ private Session session() {
+        ClusterHealthRequest request = new ClusterHealthRequest(indices)
+            .waitForNoRelocatingShards(true)
+            .waitForActiveShards(ActiveShardCount.DEFAULT);
+        transportClusterHealthAction.execute(request, listener);
+        return listener;
+    }
+
+    private CompletableFuture<Long> initRevertAllocation(String... indices) {
+        Settings settings = Settings.builder()
+            .putNull(INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name")
+            .build();
+        return updateSettings(settings, indices);
+    }
+
+    private CompletableFuture<Long> shrinkIndex(String sourceIndexName,
+                                                @Nullable String sourceIndexAlias,
+                                                String targetIndexName,
+                                                int targetNumberOfShards) {
+        Settings targetIndexSettings = Settings.builder()
+            .put(SETTING_NUMBER_OF_SHARDS, targetNumberOfShards)
+            .build();
+
+        ResizeRequest request = new ResizeRequest(targetIndexName, sourceIndexName);
+        request.getTargetIndexRequest().settings(targetIndexSettings);
+        if (sourceIndexAlias != null) {
+            request.getTargetIndexRequest().alias(new Alias(sourceIndexAlias));
+        }
+        request.setResizeType(ResizeType.SHRINK);
+        request.setCopySettings(Boolean.TRUE);
+        request.setWaitForActiveShards(ActiveShardCount.ONE);
+
+        FutureActionListener<ResizeResponse, Long> listener
+            = new FutureActionListener<>(response -> {
+                if (response.isAcknowledged()) {
+                    activeShardsObserver.waitForActiveShards(request.getTargetIndexRequest().indices(),
+                        ActiveShardCount.ONE,
+                        request.ackTimeout(),
+                        shardsAcked -> {
+                            if (!shardsAcked) {
+                                throw new RuntimeException("Shrinking operation was successful but the operation " +
+                                                           "timed out while waiting for enough shards to be started.");
+                            }
+                        },
+                        e -> {
+                            throw new IllegalStateException(e);
+                        });
+                } else {
+                    throw new RuntimeException("Publishing new cluster state during Shrink operation (shrink phase) " +
+                                               "has timed out.");
+                }
+                return 0L;
+            });
+
+        transportResizeAction.execute(request, listener);
+        return listener;
+    }
+
+    private CompletableFuture<Long> deleteIndex(String indexName) {
+        DeleteIndexRequest request = new DeleteIndexRequest(indexName);
+
+        FutureActionListener<AcknowledgedResponse, Long> listener = new FutureActionListener<>(r -> 0L);
```

Comment:
do we care about a possible failure of the delete operation? ( I'm thinking of deleting the backup as for the shrinked leftover we have an operation to clean up)

Reply:
if the operation is retried, the backup is also deleted before the new shrink begins.
In any case, the backup concept will go in #7992